environment.yml

```diff
@@ -1,38 +1,35 @@
-name: liionpack
+# Use this config file to create a conda development environment. This
+# environment installs liionpack in editable mode which is useful for
+# development of the liionpack source code. General users should install
+# liionpack via pip.
+
+# Create a conda environment named lipack
+# $ cd liionpack
+# $ conda env create --file environment.yml
+
+# Activate the environment
+# $ conda activate lipack
+
+# Exit the environment
+# $ conda deactivate
+
+# Delete the environment
+# $ conda env remove --name lipack
+
+name: lipack
+
 dependencies:
-  - python
-  - pip
+  - python=3.9.*
   - numpy
   - scipy
   - matplotlib
   - pandas
-  - plotly
-  - openpyxl
-  - tqdm
-  - networkx
-  - jupyter
-  - sympy
   - ipython
-  - wrapt
+  - jupyter
   - sympy
-  - redis
-  - pyyaml
-  - protobuf >=3.8.0,<4.0.0
-  - imageio
-  - grpcio
-  - future
-  - flatbuffers
-  - filelock
-  - deprecated
-  - click
-  - autograd
-  - absl-py
-  - hiredis
+  - flake8
+  - pip
   - pip:
-    - pybamm
-    - lcapy
-    - scikit-spatial
-    - ray
-    - mkdocstrings-python-legacy
-    - mkdocs-material
-    - mkdocs-jupyter
\ No newline at end of file
+    - pybamm==22.6
```

Comment:
Out of curiosity, is there a reason for pinning `pybamm` so tightly? Adding upper caps to dependencies usually creates a lot of mess. A very good read on this - https://iscinumpy.dev/post/bound-version-constraints/

Reply:
I keep having issues when I update liionpack because of breaking changes introduced in PyBaMM releases. So I pinned the PyBaMM version for the conda environment to avoid this problem.

Reply:
IMO it should be okay if the development version of `liionpack` breaks due to a breaking change in `pybamm`, as that can be fixed without disturbing the users. Hard capping would make `liionpack` behave as an application instead of a library.

For instance, let's say that a user has a library that depends on `liionpack`. Now, if we make a new release `v0.x` with `pybamm==22.6`, and the user had the dependencies defined as the following -
```
pybamm>=22.7
liionpack
```
then the user will never be able to access `liionpack v0.x`, as modern package managers would automatically pick up the last version of `liionpack` which is compatible with `pybamm>=22.7` (in this case, the version released before `0.x`).

In the long run, this would create an issue, and this is just one of the many issues associated with hard capping dependencies.

Another interesting take on this is that breaking changes are relative. One user's bug fix can be another user's breaking change. For instance, take a look at https://github.com/numpy/numpy/issues/20929. The user was using a bug in `NumPy's codebase as a feature, and their project's CI went down once the bug was fixed.

But, `liionpack` is not available through `conda` yet (if I am not wrong), so maybe users (not developers) could be discouraged to use the `conda` installation? Then we can remove this hard cap when `liionpack` is released to `conda`! Up to you and @TomTranter! :)

Reply:
There have been a few things recently with some of the pybamm and liionpack dependencies themselves requiring specific versions of other packages. sympy and protobuf in particular had version conflicts. I hope these are all sorted now and will be bound to happen from time to time. In general I don't like pinning versions and as we maintain both liionpack and pybamm we should be able to make sure they both work together on latest versions. That's my view. If we want to maintain a permanent dev branch with pinned versions so that it just works and then merge to master when things get ironed out elsewhere that could work maybe.

Reply:
@Saransh-cpp would it be possible to maintain a dev branch and a main branch and have dev point to pybamm dev so we can catch breaking changes early?

Reply:
Yes! But will this work as `PyBaMM`'s develop and main branch? That is pushing to main only when we are releasing a new version of `liionpack` and merging the regular pull requests in `develop`?

Reply:
If this is too much work then don't worry about it. I can make sure I manually update PyBaMM when a new version is released.

Reply:
I have pushed a `develop` branch, but that can be abandoned if it looks like this would require extra time for maintenance. But, the workflows should be changed somehow, as we are testing `liionpack` with the latest `PyBaMM` commit, but we are releasing `liionpack` with the latest tagged version of `PyBaMM`.

Reply:
@wigging would you be able to apply these changes to a new PR with the develop branch as the base and remove the version pin for pybamm

Reply:
I changed the base branch and PyBaMM version. Had to do some rebases and force pushes so hopefully I didn't screw up anything.